### ISSUE TITLE
The mime type does not match to file extension

### ISSUE THREAD

**Issue by fampla** [X-Ryl669/grav-plugin-viewer3d#5]
hello,
 I probably have a DAU question but I keep getting the error.
"The mime type does not match to file extension"
I entered it in the media and went to save,
<img width="1367" alt="Bildschirmfoto 2021-07-22 um 10 44 28" src="https://user-images.githubusercontent.com/22618979/126612365-4d452ac5-60e6-4481-bdb6-999a52ef0007.png">

 but Grav does not want to upload an .stl file.
what am I doing wrong?

**Comment by slevin-7** [comment]
okay, if I don't upload the files to the folder via the admin panel, everything works without any problems.
Maybe that was exactly my problem, that I didn't copy it into the folder straight away.

**Comment by X-Ryl669** [comment]
Strange, I have the same configuration and it works for me. Did you reload the admin panel (and delete all cookies) after changing the *media* tab?

**Comment by X-Ryl669** [comment]
I'm not able to reproduce it. I'm able to upload STL file directly from the admin panel's page edition tool. If you think it still happens for you, I'll need more information to understand why it doesn't work for you. Thanks!

**Comment by khcassidy** [comment]
Concerning the error: My Mime type had to be changed to:
  stl:
    type: file
    thumb: media/thumb-stl.png
    mime: application/octet-stream
    image: null

Check your stl file type here ... https://www.htmlstrip.com/mime-file-type-checker

**Comment by X-Ryl669** [comment]
I don't recommend using `application/octet-stream` for the mime type, since it's a "don't know what it is" mime type. So, when you set up this way, a STL file is considered as a bunch of bytes without any meaning. On some platform (typically Windows), the system will use the extension to figure out the type, so it might appear to work, but it's not ok for other systems (linux, mac, ios, etc...). 

The correct mime type for STL is `model/stl` as registered by [IANA here](https://www.iana.org/assignments/media-types/model/stl)

**Comment by khcassidy** [comment]
Yup, I agree it should be model/stl but .... but I could not get it working any other way.
on my linux web server and laptop I run:

\# file --mime-type ThinkLightv1.stl
ThinkLightv1.stl: application/octet-stream

My Web Server itself has stl associated to the model/stl mime type as does Apache. Maybe its wrong somewhere but I can not find where to fix it on my systems. Maybe its Freecad export issue? Either way its not a plugin issue and I am extremely thankful for this plugin, and works perfectly for me.

**Comment by X-Ryl669** [comment]
Well, it might be the file's database that's wrong, but it's not important here, because the mime type to use in the configuration page is the one reported by the web browser and not the OS. When you upload a file in the grav's page editor, it's the browser that detect the mime type and try to upload it to the server. grav's code only compare what the browser detects with what is allowed, and if allowed, proceed to the upload (else it errors out).

To figure out what is this mime type, you can use the [demo here](https://developer.mozilla.org/en-US/docs/Web/API/File/type), don't worry it's local only. Select a stl file and it should report "model/stl". If it doesn't well, I guess we need a fallback route. But in that case, I would report the error to the browser's bug tracker.

Also, there are 2 kinds of STL file: binary STL and ascii STL. Both use .stl extension and should both report `model/stl` as a mime-type.
